### PR TITLE
Project model enhancement

### PR DESCRIPTION
**Pull request by KaiwalyaPataskar**

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -44,6 +44,9 @@ def create
   end
 
   def update
+    if @project.repositories.count > 0 && params[:project].keys.include?('repositories_attributes')
```
Comment (anujaware) [inline app/controllers/projects_controller.rb:47]: @KaiwalyaPataskar Why do we need this?

File: lib/tasks/copy_project_repo_details_to_repositories.rake [hunk 2]
```diff
@@ -0,0 +1,21 @@
+namespace :create_project_repositories do
+  desc "Copies CodeClimate details of all Projects to Repository model."
+  task :copy_code_climate_details => [:environment] do
+    projects = Project.all
+    repositories = Repository.count
+    projects.each do |project|
+      if project.repositories.empty? &&
```
Comment (anujaware) [inline lib/tasks/copy_project_repo_details_to_repositories.rake:7]: @KaiwalyaPataskar We can have a button in repo details page 'Fetch Code Climate Details'. On click will run following code snippet for linking.

File: app/models/repository.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  HOSTS = ['GitHub', 'BitBucket', 'GitLab', 'SourceForge', 'Launchpad', 'Google Cloud', 'AWS CodeCommit'].freeze
+  belongs_to :project
+  field :name
+  field :host
+  field :url
+  field :code_climate_id
+  field :maintainability_badge
+  field :test_coverage_badge
+  # validates_presence_of :name, :url, :host
+  # validates :host, inclusion: { in: HOSTS, allow_nil: false }
```
Comment (KaiwalyaPataskar) [inline app/models/repository.rb:12]: We'll need to uncomment above lines after we run **copy_project_repo_details_to_repositories** task.

File: config/initializers/simple_form.rb [hunk 4]
```diff
@@ -111,7 +111,7 @@
   # in this configuration, which is recommended due to some quirks from different browsers.
   # To stop SimpleForm from generating the novalidate option, enabling the HTML5 validations,
   # change this configuration to true.
-  config.browser_validations = false
+  config.browser_validations = true
```
Comment (swapnilchincholkar) [inline config/initializers/simple_form.rb:114]: @KaiwalyaPataskar Changing global configs can impact entire existing application, please ensure to thoroughly test the effects caused by such changes.

Comment (KaiwalyaPataskar) [reply config/initializers/simple_form.rb]: Disabled it back, as we've handled validations in forms.

File: app/controllers/projects_controller.rb (hunk 1 above)
Comment (KaiwalyaPataskar) [reply app/controllers/projects_controller.rb]: Removed it, as it is getting handled properly by nested_form.

File: lib/tasks/weekly_code_climate_statistics.rake [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+              stats[metric] = point["value"] || 0
+            end
+            stats[:timestamp] = Time.at(point["timestamp"]) if !stats[:timestamp]
+          end
+        end
+        code_climate_stat = CodeClimateStatistic.create(stats)
+        if code_climate_stat.valid?
+          puts "CodeClimate Statistics created successfully for #{repo.project.name}\'s #{repo.project.name} Repository."
+        else
+          puts "Error: #{code_climate_stat.errors.full_messages.join(",")} for #{repo.project.name}"
```
Comment (anujaware) [inline lib/tasks/weekly_code_climate_statistics.rake:40]: @KaiwalyaPataskar As are running this script in the background we need to capture in db against repository.

Comment (KaiwalyaPataskar) [reply lib/tasks/weekly_code_climate_statistics.rake:40]: Let's create a new Issue for this, as it needs to be addressed generically.

File: lib/tasks/weekly_code_climate_statistics.rake [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  to = Date.today.strftime('%Y-%m-%d')
+  query_string = "filter[from]=#{from}&filter[to]=#{to}"
+  Repository.each do |repo|
+    if repo.code_climate_id && !repo.code_climate_id.empty?
+      url = "https://api.codeclimate.com/v1/repos/#{repo.code_climate_id}/metrics?#{query_string}"
+      headers = { "Accept" => "application/vnd.api+json", "Authorization" => "Token token=#{ENV['CODE_CLIMATE_TOKEN']}" }
+      begin
+        response = HTTParty.get(url, headers: headers, timeout: 20)
+      rescue Timeout::Error => e
+        puts "Error: Request Timeout for #{repo.project.name}"
```
[the remark this answers is not in the document]
Comment (anujaware) [reply lib/tasks/weekly_code_climate_statistics.rake:14]: Added #366

**Comment by anujaware** [comment]
@swapnilchincholkar Can you please merge this PR into `master` and deploy on staging for testing?
After deployment we need to run a task to copy data from `Project` model to `Repository` model.

**Pull request merged**